### PR TITLE
[RFC/WIP] [Still failing tests] Better invariant - leaf guards are always hasattr/item protected

### PR DESCRIPTION
**Pull request by voznesenskym**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110772

There is a rare state we can get into - wherein we produce leaf guards for objects without properly protecting the path to them via checks.

For a pathological example found by @Fidget-Spinner  - a functools.partial created on one side of a graph break with an integer value in its keywords does not have its guards installed until an invocation on the other side of the graph break. However, due to how we handle scalars, that keyword arguments value was lifted as an input. This means that the first graph has a guard for value like:

`op.keywords['dim']`

This is sound unless the functools partial, op, does not match the same state on future invocations. The guards won't just fail, they will crash.

The correct thing to do is to add a guard before this check, so that the dim key access is protected.

now, we can do this manually, trivially, with functools.partials - but I worry about the brittleness of our infra that encountering a leaf guard (TYPE_MATCH in this case) does not automatically guarantee that you accumulate all the necessary guards in the chain to that leaf.

I think the guards C++ rewrite that @anijain2305  will jam on when he gets back will solve this, but in the interim, I jotted out this PR as a quick implementation, but there are other ways of doing this. Looking for feedback. 

cc @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

**Comment by anijain2305** [comment]
Is `VariableTracker.propagate` a better place to put this?

File: torch/_dynamo/variables/builder.py [hunk 1]
```diff
@@ -1184,7 +1186,7 @@ def wrap_unspecialized_primitive(self, value):
 
                 name = self.source.name()
                 if name not in self.tx.output.frame_state:
-                    # Note - this essentially means that if this name gets reused as a tensor,
+                    # Note - this esentially means that if this name gets reused as a tensor,
```
Comment (ezyang) [inline torch/_dynamo/variables/builder.py:1189]: lol

**Comment by ezyang** [comment]
It's not obvious from the PR desc, nor a quick skim of the PR, how I'm supposed to write guard code differently in Dynamo post this PR.

**Review by jansel — CHANGES_REQUESTED** [review]
I think going with @anijain2305's C++ guards structure is a better solution to this.  

In a lot of cases these added guards are redundant with:
- TYPE_MATCH guards for things like range/slice
- DICT_KEYS guards for dictionaries
- etc

I think an easier short term fix is just to add the missing guard for partial.

File: torch/_dynamo/variables/builder.py (hunk 1 above)
Comment (voznesenskym) [reply torch/_dynamo/variables/builder.py:1189]: I think we added a spellchecker to the linter - and it did it wrong lol

**Comment by voznesenskym** [comment]
> It's not obvious from the PR desc, nor a quick skim of the PR, how I'm supposed to write guard code differently in Dynamo post this PR.

You are not.

**Comment by voznesenskym** [comment]
> I think going with @anijain2305's C++ guards structure is a better solution to this.

Agreed

> In a lot of cases these added guards are redundant with:
> 
> * TYPE_MATCH guards for things like range/slice
> * DICT_KEYS guards for dictionaries
> * etc
> 

How can guards be redundant? they are a set. If I add a TYPE_MATCH guard to a dict key, it adds as HAS_ITEM guard to that dict key, which then becomes a DICT_KEYS guard on the dict.

**Comment by ezyang** [comment]
@jansel while guard trees will help with getting correct ordering, we still need to install the right guards, which won't happen automatically, I think?

**Comment by jansel** [comment]
> @jansel while guard trees will help with getting correct ordering, we still need to install the right guards, which won't happen automatically, I think?

The existence checks would go away and just become part of the traversal.

**Review by jansel — CHANGES_REQUESTED** [review]
Let's just add the missing guard to `functools.partial`.

If we wanted to be sloppy and stop worrying about correctly guarding attribute access, we could just catch and ignore `AttributeError`/`KeyError` when running guards.

**Comment by anijain2305** [comment]
> @jansel while guard trees will help with getting correct ordering, we still need to install the right guards, which won't happen automatically, I think?

These can happen automatically in Accessors in guard trees refactor. For each accessor, we can check if the accessor can even access the value. I think we can even remove guards like DICT_KEYS, because the accessor is more fine grained.

**Comment by anijain2305** [comment]
For this PR, another idea is to add an extra guard `getattr(a, b, False)` whenever we are adding any guard if the guard.originating source is AttrSource (and similarly for GetItemSource). CSE part should take care of deduping it.

**Comment by voznesenskym** [comment]
> > @jansel while guard trees will help with getting correct ordering, we still need to install the right guards, which won't happen automatically, I think?
> 
> The existence checks would go away and just become part of the traversal.
> 
> I don't think there is any shortcut to getting the guards right.

Thats what this PR does? I don't understand the pushback here. I am happy to account for anything I missed, but so far it feels ground in platitudes. None of the initially raised concerns are correct. This does not create duplicate guards for dicts, for example, as stated.

**Comment by voznesenskym** [comment]
> For this PR, another idea is to add an extra guard `getattr(a, b, False)` in all the guard create_functions if the guard.originating source is AttrSource (and similarly for GetItemSource). CSE part should take care of deduping it (hmm, maybe not, CSE might be only for sources and not guard exprs).

This just betrays soundness guarantees and adds overhead to each guard. We should just accumulate the right thing (What this PR adds)

**Comment by jansel** [comment]
If we want to go with this approach then we should just catch and ignore `AttributeError`/`KeyError` when running guards.  This will be faster and a less code and does the same thing as this PR.

The extra guards is real:
- TYPE_MATCH on `functools.partial` implies `.args`, `.func`, `.keywords` exist, you don't need extra guards for them.
- TYPE_MATCH on `slice` implies `.start`, `.step`, `.stop` exist, you don't need extra guards for them
- `.__class__` does not need extra guarding
- Function types don't need checking for `__closure__`/`__code__`/etc

Blinding adding guards on every usage is not correct.

**Comment by voznesenskym** [comment]
> If we want to go with this approach then we should just catch and ignore `AttributeError`/`KeyError` when running guards. This will be faster and a less code and does the same thing as this PR.
> 
> The extra guards is real:
> 
> * TYPE_MATCH on `functools.partial` implies `.args`, `.func`, `.keywords` exist, you don't need extra guards for them.
> * TYPE_MATCH on `slice` implies `.start`, `.step`, `.stop` exist, you don't need extra guards for them
> * `.__class__` does not need extra guarding
> * Function types don't need checking for `__closure__`/`__code__`/etc
> 
> Blinding adding guards on every usage is not correct.

These could all be routed to their correct respective path in HAS_ITEM - or dropped entirely. It's not blindly adding guards on every usage. 

> See failing tests for lots more issues.

This just means we didn't cover every edge case.

The fact of the matter is we have a bug in the architecture, and the functools.partial case just exercised it. 

I don't care enough about this to argue further though. We can just hack fix the partials case and move on. 

Hack fix for partials here: https://github.com/pytorch/pytorch/pull/110802

**Pull request closed without merging**